sdk/core/src/error/mod.rs

```diff
@@ -0,0 +1,371 @@
+        }
+    }
+
+    /// Consumes the error attempting to downcast the inner error as the type provided
+    ///
+    /// Returns `Err(self)` if the downcast is not possible
+    pub fn into_downcast<T: std::error::Error + 'static>(self) -> std::result::Result<T, Self> {
+        if self.downcast_ref::<T>().is_none() {
+            return Err(self);
+        }
+        // Unwrapping is ok here since we already check above that the downcast will work
+        Ok(*self
+            .into_inner()?
+            .downcast()
+            .expect("failed to unwrap downcast"))
+    }
+
+    /// Returns a reference to the inner error wrapped by this error (if any).
+    pub fn get_ref(&self) -> Option<&(dyn std::error::Error + Send + Sync + 'static)> {
+        match &self.context {
+            Context::Custom(Custom { error, .. }) => Some(error.as_ref()),
+            Context::Full(Custom { error, .. }, _) => Some(error.as_ref()),
+            _ => None,
+        }
+    }
+
+    /// Returns a reference to the inner error (if any) downcasted to the type provided
+    pub fn downcast_ref<T: std::error::Error + 'static>(&self) -> Option<&T> {
+        self.get_ref()?.downcast_ref()
+    }
+
+    /// Returns a mutable reference to the inner error wrapped by this error (if any).
+    pub fn get_mut(&mut self) -> Option<&mut (dyn std::error::Error + Send + Sync + 'static)> {
+        match &mut self.context {
+            Context::Custom(Custom { error, .. }) => Some(error.as_mut()),
+            Context::Full(Custom { error, .. }, _) => Some(error.as_mut()),
+            _ => None,
+        }
+    }
+
+    /// Returns a mutable reference to the inner error (if any) downcasted to the type provided
+    pub fn downcast_mut<T: std::error::Error + 'static>(&mut self) -> Option<&mut T> {
+        self.get_mut()?.downcast_mut()
+    }
+}
+
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match &self.context {
+            Context::Custom(Custom { error, .. }) => error.source(),
+            Context::Full(Custom { error, .. }, _) => error.source(),
+            _ => None,
+        }
+    }
+}
+
+impl From<ErrorKind> for Error {
+    fn from(kind: ErrorKind) -> Self {
+        Self {
+            context: Context::Simple(kind),
+        }
+    }
+}
+
+impl From<std::io::Error> for Error {
+    fn from(error: std::io::Error) -> Self {
+        Self {
+            context: Context::Custom(Custom {
+                kind: ErrorKind::Io,
+                error: Box::new(error),
+            }),
+        }
+    }
+}
+
+impl From<serde_json::Error> for Error {
+    fn from(error: serde_json::Error) -> Self {
+        Self {
+            context: Context::Custom(Custom {
+                kind: ErrorKind::DataConversion,
+                error: Box::new(error),
+            }),
+        }
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```

Comment:
For `HttpResponse` we actually do tend to format sanitized responses like so:

```
Message: 
    Azure.RequestFailedException : Service request failed.
    Status: 404 (Not Found) 
Content:
    {"error":{"code":"SecretNotFound","message":"Secret not found: some_secret"}}
    
Headers:
    Cache-Control: no-cache
    Pragma: no-cache
    Server: Microsoft-IIS/10.0
    x-ms-keyvault-region: westus
    x-ms-request-id: 625f870e-10ea-41e5-8380-282e5cf768f2
    x-ms-keyvault-service-version: 1.1.0.866
    x-ms-keyvault-network-info: addr=131.107.174.199;act_addr_fam=InterNetwork;
    X-AspNet-Version: 4.0.30319
    X-Powered-By: ASP.NET
    Strict-Transport-Security: max-age=31536000;includeSubDomains
    X-Content-Type-Options: nosniff
    Date: Tue, 18 Jun 2019 16:02:11 GMT
    Content-Length: 75
    Content-Type: application/json; charset=utf-8
    Expires: -1
```

While the format may vary a bit from language to language SDK, the information displayed by default is fairly consistent. Most errors developers will see with the Azure SDKs are service responses, and (sanitized) responses such as these make diagnostics much easier - even in production / logs.

Reply:
I've added a more interesting display implementation that mirrors what you have above. We can keep improving it over time.